### PR TITLE
Add native type declarations

### PR DESCRIPTION
**Pull request by othercorey**

### CODE REVIEW

File: src/Authenticator/AuthenticatorCollection.php [hunk 1]
```diff
@@ -49,19 +49,21 @@ public function __construct(IdentifierCollection $identifiers, array $config = [
...
      *
-     * @param string $class Authenticator class.
+     * @param object|string $class Authenticator class.
      * @param string $alias Authenticator alias.
      * @param array $config Config array.
      * @return \Authentication\Authenticator\AuthenticatorInterface
      * @throws \RuntimeException
      */
-    protected function _create($class, string $alias, array $config): AuthenticatorInterface
+    protected function _create(object|string $class, string $alias, array $config): AuthenticatorInterface
```
Comment (othercorey) [inline src/Authenticator/AuthenticatorCollection.php:58]: We must support `object|string` to implement ObjectRegistry.

File: src/Identifier/IdentifierCollection.php [hunk 2]
```diff
@@ -66,19 +67,21 @@ public function identify(array $credentials)
...
      *
-     * @param string $class Identifier class.
+     * @param object|string $class Identifier class.
      * @param string $alias Identifier alias.
      * @param array $config Config array.
      * @return \Authentication\Identifier\IdentifierInterface
      * @throws \RuntimeException
      */
-    protected function _create($class, string $alias, array $config): IdentifierInterface
+    protected function _create(object|string $class, string $alias, array $config): IdentifierInterface
```
Comment (othercorey) [inline src/Identifier/IdentifierCollection.php:76]: We must support `object|string` to implement ObjectRegistry.

File: src/PasswordHasher/DefaultPasswordHasher.php [hunk 3]
```diff
@@ -40,10 +40,10 @@ class DefaultPasswordHasher extends AbstractPasswordHasher
     /**
      * Generates password hash.
      *
-     * @param string $password Plain text password to hash.
+     * @param array|string $password Plain text password to hash.
      * @return string Password hash or false on failure.
      */
-    public function hash($password): string
+    public function hash(array|string $password): string
```
Comment (othercorey) [inline src/PasswordHasher/DefaultPasswordHasher.php]: PasswordHasherInterface requires `array|string` for password, but none of the hashers support arrays.

Comment (ndm2) [reply src/PasswordHasher/DefaultPasswordHasher.php]: That doesn't sound right. Where exactly does the identifier pass an array for the password?

Comment (othercorey) [reply src/PasswordHasher/DefaultPasswordHasher.php]: It was just part of the @param annotation that was converted into a type declaration.

Comment (ndm2) [reply src/PasswordHasher/DefaultPasswordHasher.php]: Sorry, I still don't see it. Which `@param`?

Comment (othercorey) [reply src/PasswordHasher/DefaultPasswordHasher.php]: https://github.com/cakephp/authentication/blob/cake5/src/PasswordHasher/PasswordHasherInterface.php#L26

Comment (ADmad) [reply src/PasswordHasher/DefaultPasswordHasher.php]: I can't remember either why it's `string|array`. Maybe to support some edge case custom hashing is my guess.

Comment (ndm2) [reply src/PasswordHasher/DefaultPasswordHasher.php]: Ah OK, you meant the interface, I thought the typo was meant to be `PasswordIdentifier`.

Comment (ndm2) [reply src/PasswordHasher/DefaultPasswordHasher.php]: Maybe this is a chance to drop the array? With parameter type widening, one could implement a `string` only param as `array|string`, or even drop the type hint completely if needed.

The password identifier doesn't support passing arrays, so a custom identifier would be needed anyways if people wanted to pass arrays to the hasher, so they could then also "cast" the used hasher accordingly.

Comment (othercorey) [reply src/PasswordHasher/DefaultPasswordHasher.php]: So we want to drop `array`? That would fix the mismatch between needsRehash() and hash().

Comment (ndm2) [reply src/PasswordHasher/DefaultPasswordHasher.php]: I don't really see a mismatch there, as `needsRehash()` operates on the hashed password (maybe the argument should be renamed to reflect that better). What makes me feel uneasy is forcing all the hashers to accept an array when they don't support it, and will fail with type errors accordingly 🤷

Comment (othercorey) [reply src/PasswordHasher/DefaultPasswordHasher.php]: ah, you're right, it takes the hashed password. I must have been looking at a different password_* parameter.

**Review by ADmad — APPROVED** [review]

**Review by markstory — APPROVED** [review]
Looks good to me.

File: src/PasswordHasher/PasswordHasherInterface.php [hunk 4]
```diff
@@ -23,21 +23,21 @@ interface PasswordHasherInterface
     /**
      * Generates password hash.
      *
-     * @param string|array $password Plain text password to hash or array of data
+     * @param string $password Plain text password to hash or array of data
      *   required to generate password hash.
```
Comment (ndm2) [inline src/PasswordHasher/PasswordHasherInterface.php]:
```suggestion
     * @param string $password Plain text password to hash.
```

File: src/PasswordHasher/PasswordHasherInterface.php [hunk 5]
```diff
@@ -23,21 +23,21 @@ interface PasswordHasherInterface
...
      */
-    public function hash($password): string;
+    public function hash(string $password): string;
 
     /**
      * Check hash. Generate hash from user provided password string or data array
      * and check against existing hash.
      *
-     * @param string|array $password Plain text password to hash or data array.
+     * @param string $password Plain text password to hash or data array.
```
Comment (ndm2) [inline src/PasswordHasher/PasswordHasherInterface.php]:
```suggestion
     * @param string $password Plain text password to hash.
```

**Pull request merged**